### ISSUE TITLE
secure box authentication URL

### ISSUE THREAD

**Issue by jonjacksonma** [rclone/rclone#3573]
#### What is the problem you are having with rclone?

Box authentication is not secure in a multi-user environment

A situation occurred where UserA connected to a UserB's rclone config instance via the http://127.0.0.1:53682/auth URL, because the URL and port are always the same and there is no mechanism to verify which user is connecting.  UserA was not aware they had connected to someone else's authentication page and went ahead to authenticate.  The secret token for UserA was passed to UserB's rclone instance and saved in UserB's rclone configuration.  This scenario results in data exposure.  I'm filing it under bugs rather than feature requests due to the security implications.

Would it be possible to embed a long random string into the auth URL and only accept connections that are using the correct random string?  Another improvement would be to pick the next available port if 53682 is taken.

#### What is your rclone version (output from `rclone version`)

rclone v1.49.3-100-g6750af61-beta
- os/arch: linux/amd64
- go version: go1.13.1

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Linux 64 bit

####  Which cloud storage system are you using? (eg Google Drive)

Box

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

rclone config

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

CONFIGURING A BOX REMOTE:
```
Use auto config?
 * Say Y if not sure
 * Say N if you are working on a remote or headless machine
y) Yes
n) No
y/n> y
If your browser doesn't open automatically go to the following link: http://127.0.0.1:53682/auth
Log in and authorize rclone for access
Waiting for code...
```

The problem / bug is that http://127.0.0.1:53682/auth is not unique - any user on a shared platform can connect

**Comment by ncw** [comment]
> The problem / bug is that http://127.0.0.1:53682/auth is not unique - any user on a shared platform can connect

Unfortunately that URL needs to be configured at box as the redirection URL so it isn't possible to change it on the fly.

I managed to replicate this.

I notice that when you start the second rclone and get to the webserver prompt you get this and rclone stops

```
Use auto config?
 * Say Y if not sure
 * Say N if you are working on a remote or headless machine
y) Yes
n) No
y/n> y
2019/09/27 13:52:00 Failed to start auth webserver: listen tcp 127.0.0.1:53682: bind: address already in use
```

However rclone has already opened the browser window at that point and popped it to the front.

So I propose two fixes

1. open the browser window after we have successfully started the oauth server
    - this would have fixed the problem on its own
2. check the unique state in the auth redirect is correct
    - this would also have fixed the problem on its own

Can you have a go with this fix and see if it fixes it for you?

https://beta.rclone.org/branch/v1.49.3-101-gda77b2d0-fix-3573-auth-beta/ (uploaded in 15-30 mins)

**Comment by jonjacksonma** [comment]
I've tested https://beta.rclone.org/branch/v1.49.3-101-gda77b2d0-fix-3573-auth-beta

It solves the problem by checking the state matches.  Will not allow any connection that doesn't provide the correct random state string.  I would consider this a fix
thanks!

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.50

I think I'll probably put this in a point release v1.49.4 also

**Comment by jonjacksonma** [comment]
That’s perfect, thanks for the quick fix.
